docs/state_encryption.md

```diff
@@ -0,0 +1,333 @@
+      }
+    }
+  },
+  "method": {
+    "aes_gcm": {
+      "foo": {
+        "key_provider": "${key_provider.static.my_key}"
+      }
+    }
+  },
+  "statefile": {
+    "method": "${method.aes_gcm.foo}"
+  }
+}
+```
+
+The user can set either of these structures in the `TF_ENCRYPTION` environment variable:
+
+```bash
+export TF_ENCRYPTION='{"key_provider":{...},"method":{...},"statefile":{...}}'
+```
+
+When the user specifies both an environment and a code configuration, `tofu` merges the two configurations. If two values conflict, the environment configuration takes precedence.
+
+To ensure that the encryption cannot be accidentally forgotten or disabled and the data stored unencrypted, the user can specify the `enforced` option in the HCL configuration:
+
+```hcl2
+terraform {
+  encryption {
+    //...
+    statefile {
+      enforced = true
+    }
+    planfile {
+      enforced = true
+    }
+    backend {
+      enforced = true
+    }
+    terraform_remote_states {
+      default {
+        enforced = true
+      }
+    }
+  }
+}
+```
+
+**Note:** The `enforced` option is also available in the environment configuration and works as intended, but doesn't make much sense because its primary purpose is to guard against environment variable omission.
+
+## Encrypted state format
+
+When `tofu` encrypts a state file, the encrypted state is still a JSON file. Any implementations can distinguish encrypted files by the `encryption` key being present in the JSON structure. However, there may be other keys in the state file depending on the encryption method and type.
+
+```json
+{
+  "encryption": "some data here"
+  // ... Additional keys here
+}
+```
+
+## Encrypted plan format
+
+A unencrypted plan file in OpenTofu is a ZIP file. This specification makes no rules for how the encrypted format should look like and all non-encryption routines should treat the value as opaque. For ease of use and recovery, however, the implementation may want to consider using a standard ZIP header to encode the details.
+
+## Implementation
+
+When implementing the encryption tooling, the implementation should be split in two parts: the library and the OpenTofu implementation. The library should rely on the cty types and hcl as a means to specify schema, but should not be otherwise tied to the OpenTofu codebase. This is necessary to enable encryption capabilities for third party tooling that may need to work with state and plan files.
+
+### Library implementation
+
+The encryption library should create an interface that other projects and OpenTofu itself can use to encrypt and decrypt state. The library should express its schema needs (e.g. for key provider config) using [cty](https://github.com/zclconf/go-cty) and [hcl](https://github.com/hashicorp/hcl), but be otherwise independent of the OpenTofu codebase. Ideally, the OpenTofu project should provide this library as a standalone dependency that does not pull in the entire OpenTofu dependency tree.
```

Comment:
Should the library include the functionality to parse a configuration in HCL and JSON from a string value into cty? This would free users of the library from the (IMHO very complex) interface of go-cty if all they want to do is decrypt some state file.